### PR TITLE
Add launch, update and cancel roles

### PR DESCRIPTION
**Pull request by sean-m-sullivan**
### What does this PR do?
Adds launch, update, and cancel roles to be in line with available tower modules.

### How should this be tested?
Automated tests are updated to use the new roles. Also the Automated tests are updated to work with changes in awx.

### Is there a relevant Issue open for this?
#131

### CODE REVIEW

File: roles/ad_hoc_command/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_ad_hoc_command_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/ad_hoc_command/README.md]: This line should be describing `tower_ad_hoc_commands`

File: roles/ad_hoc_command/README.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.ad_hoc_command, when: tower_ad_hoc_commands is defined}
+
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/ad_hoc_command/README.md]: Could you fill this in please

File: roles/ad_hoc_command_cancel/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_ad_hoc_command_cancel_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/ad_hoc_command_cancel/README.md]: this should be for `tower_ad_hoc_commands_cancel `

File: roles/ad_hoc_command_cancel/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        ignore_files: [tower_config.yml.template]
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.ad_hoc_command_cancel, when: tower_ad_hoc_commands is defined}
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/ad_hoc_command_cancel/README.md]: Can you fill this in please

File: roles/ad_hoc_command_cancel/tests/collections/requirements.yml [hunk 5]
```diff
@@ -0,0 +1,4 @@
+---
+collections:
+  - name: awx.awx
+    version: 11.2.0
```
Comment (Tompage1994) [inline roles/ad_hoc_command_cancel/tests/collections/requirements.yml]: Should this really be defined here? Not sure whether we do in other places and will make it harder to maintain going forwards

File: roles/inventory_source_update/README.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_inventory_source_update_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/inventory_source_update/README.md]: Should be for `tower_inventory_sources`

File: roles/inventory_source_update/README.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.inventory_source_update, when: tower_ad_hoc_commands is defined}
+
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/inventory_source_update/README.md]: Fill in please

File: roles/job_launch/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_job_launch_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/job_launch/README.md]: Should be `tower_launch_jobs`

File: roles/job_launch/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+### Secure Logging Variables
+The following Variables compliment each other.
+If Both variables are not set, secure logging defaults to false.
+The role defaults to False as normally the add ******* task does not include sensitive information.
```
Comment (Tompage1994) [inline roles/job_launch/README.md]: Can you fill in these to be relevant to the role

File: roles/job_launch/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.job_launch, when: tower_ad_hoc_commands is defined}
+
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/job_launch/README.md]: And fill in this again

File: roles/jobs_cancel/README.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_jobs_cancel_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/jobs_cancel/README.md]: Same issue. Again should be for `tower_cancel_jobs`

File: roles/jobs_cancel/README.md [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+### Secure Logging Variables
+The following Variables compliment each other.
+If Both variables are not set, secure logging defaults to false.
+The role defaults to False as normally the add ******* task does not include sensitive information.
+tower_configuration_*******_secure_logging defaults to the value of tower_configuration_secure_logging if it is not explicitly called. This allows for secure logging to be toggled for the entire suite of tower configuration roles with a single variable, or for the user to selectively use it.
```
Comment (Tompage1994) [inline roles/jobs_cancel/README.md]: Fill in

File: roles/jobs_cancel/README.md [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.jobs_cancel, when: tower_ad_hoc_commands is defined}
+
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/jobs_cancel/README.md]: Fill in

File: roles/jobs_cancel/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,20 @@
+---
+
+# Cancel a Tower Job
+- name: Launch a Tower Workflow
+  awx.awx.tower_job_cancel:
+    job_id:                   "{{ __jobs_cancel_item.id }}"
```
Comment (Tompage1994) [inline roles/jobs_cancel/tasks/main.yml:6]: I assume this is wrong

Comment (Tompage1994) [reply roles/jobs_cancel/tasks/main.yml:6]: Actually its not. I've only just spotted these. Are we doing this in any other roles? A quick look at a couple says we are not, so perhaps we shouldnt be straying from our standard?

File: roles/project_update/README.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+|`organization`|""|no|str|Organization the project exists in. Used for lookup only.|
+|`wait`|""|no|str|Wait for the project to complete.|
+|`interval`|""|no|str|The interval to request an update from Tower.|
+|`timeout`|""|no|str|If waiting for the job to complete this will abort after this amount of seconds.|
+
+### Standard Project Data Structure
+#### Yaml Example
+```yaml
+---
+tower_projects:
```
Comment (Tompage1994) [inline roles/project_update/README.md:46]: These items are wrong

Comment (Tompage1994) [reply roles/project_update/README.md:46]: Unless we are forcing an update using this? Not sure we necessarily want to use the tower_project item for this. Unless we specify an option such as `update: true`

File: roles/project_update/README.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.project_update, when: tower_ad_hoc_commands is defined}
+
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/project_update/README.md]: Fill in

File: roles/project_update/README.md [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_project_update_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/project_update/README.md]: Should be for `tower_projects`

File: roles/project_update/README.md [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+### Secure Logging Variables
+The following Variables compliment each other.
+If Both variables are not set, secure logging defaults to false.
+The role defaults to False as normally the add ******* task does not include sensitive information.
```
Comment (Tompage1994) [inline roles/project_update/README.md]: Fill these please

File: roles/workflow_launch/README.md [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+## Variables
+|Variable Name|Default Value|Required|Description|Example|
+|:---:|:---:|:---:|:---:|:---:|
+|`tower_state`|"present"|no|The state all objects will take unless overriden by object default|'absent'|
+|`tower_hostname`|""|yes|URL to the Ansible Tower Server.|127.0.0.1|
+|`validate_certs`|`False`|no|Whether or not to validate the Ansible Tower Server's SSL certificate.||
+|`tower_username`|""|yes|Admin User on the Ansible Tower Server.||
+|`tower_password`|""|yes|Tower Admin User's password on the Ansible Tower Server.  This should be stored in an Ansible Vault at vars/tower-secrets.yml or elsewhere and called from a parent playbook.||
+|`tower_oauthtoken`|""|yes|Tower Admin User's token on the Ansible Tower Server.  This should be stored in an Ansible Vault at or elsewhere and called from a parent playbook.||
+|`tower_configuration_workflow_launch_secure_logging`|`see below`|yes|Data structure describing your orgainzation or orgainzations Described below.||
```
Comment (Tompage1994) [inline roles/workflow_launch/README.md]: Shoudl be for tower_workflow_launch_jobs

File: roles/workflow_launch/README.md [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+### Secure Logging Variables
+The following Variables compliment each other.
+If Both variables are not set, secure logging defaults to false.
+The role defaults to False as normally the add ******* task does not include sensitive information.
+tower_configuration_*******_secure_logging defaults to the value of tower_configuration_secure_logging if it is not explicitly called. This allows for secure logging to be toggled for the entire suite of tower configuration roles with a single variable, or for the user to selectively use it.
```
Comment (Tompage1994) [inline roles/workflow_launch/README.md]: Fill these

File: roles/workflow_launch/README.md [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        extensions: ["yml"]
+  roles:
+    - {role: redhat_cop.tower_configuration.workflow_launch, when: tower_ad_hoc_commands is defined}
+
+```
+## License
+[MIT](LICENSE)
+
+## Author
+[************](************)
```
Comment (Tompage1994) [inline roles/workflow_launch/README.md]: And these

**Review by Tompage1994 — CHANGES_REQUESTED** [review]
Good work. A fair few things to correct, though most are just repeats throughout

File: roles/ad_hoc_command_cancel/tests/collections/requirements.yml (hunk 5 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command_cancel/tests/collections/requirements.yml]: Updating all of them to remove the version, think the requirements file should be there, but agree on removing version to make it easier to maintain.

File: roles/ad_hoc_command/README.md (hunk 1 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command/README.md]: Fixed and updated.

File: roles/ad_hoc_command/README.md (hunk 2 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command/README.md:90]: Fixed and updated.

File: roles/jobs_cancel/tasks/main.yml (hunk 14 above)
Comment (sean-m-sullivan) [reply roles/jobs_cancel/tasks/main.yml:6]: I started doing it on new roles for any internal variables to match the oasis standard. The basics is, if the user will not be using the variable it should be started as __ to denote an internal variable. 

We can change it to the old standard, but was thinking we should change it over collection wide to the new.

File: roles/workflow_launch/README.md (hunk 21 above)
Comment (sean-m-sullivan) [reply roles/workflow_launch/README.md]: Updated and fixed

File: roles/workflow_launch/README.md (hunk 20 above)
Comment (sean-m-sullivan) [reply roles/workflow_launch/README.md]: Updated and fixed

File: roles/project_update/README.md (hunk 16 above)
Comment (sean-m-sullivan) [reply roles/project_update/README.md]: Updated and fixed

File: roles/project_update/README.md (hunk 15 above)
Comment (sean-m-sullivan) [reply roles/project_update/README.md:46]: I actually made it so it used the tower projects variable by default so users can set which ones to be true or not in the same config file rather then keeping the data updated in two places. Or should we do it that way? They could always load a separate file with the same variable set.

File: roles/project_update/README.md (hunk 18 above)
Comment (sean-m-sullivan) [reply roles/project_update/README.md]: Updated and fixed

File: roles/project_update/README.md (hunk 17 above)
Comment (sean-m-sullivan) [reply roles/project_update/README.md]: Updated and fixed

File: roles/jobs_cancel/README.md (hunk 13 above)
Comment (sean-m-sullivan) [reply roles/jobs_cancel/README.md]: Updated and fixed

File: roles/jobs_cancel/README.md (hunk 12 above)
Comment (sean-m-sullivan) [reply roles/jobs_cancel/README.md]: Updated and fixed

File: roles/jobs_cancel/README.md (hunk 11 above)
Comment (sean-m-sullivan) [reply roles/jobs_cancel/README.md]: Updated and fixed

File: roles/job_launch/README.md (hunk 10 above)
Comment (sean-m-sullivan) [reply roles/job_launch/README.md]: Updated and fixed

File: roles/job_launch/README.md (hunk 9 above)
Comment (sean-m-sullivan) [reply roles/job_launch/README.md]: Updated and fixed

File: roles/job_launch/README.md (hunk 8 above)
Comment (sean-m-sullivan) [reply roles/job_launch/README.md]: Updated and fixed

File: roles/inventory_source_update/README.md (hunk 7 above)
Comment (sean-m-sullivan) [reply roles/inventory_source_update/README.md]: Updated and fixed

File: roles/inventory_source_update/README.md (hunk 6 above)
Comment (sean-m-sullivan) [reply roles/inventory_source_update/README.md]: Updated and fixed

File: roles/ad_hoc_command_cancel/README.md (hunk 4 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command_cancel/README.md]: Updated and fixed

File: roles/ad_hoc_command_cancel/README.md (hunk 3 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command_cancel/README.md]: Updated and fixed

File: roles/ad_hoc_command/README.md (hunk 2 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command/README.md]: Updated and fixed

File: roles/ad_hoc_command/README.md (hunk 1 above)
Comment (sean-m-sullivan) [reply roles/ad_hoc_command/README.md]: Updated and fixed

File: roles/jobs_cancel/tasks/main.yml (hunk 14 above)
Comment (kedark3) [reply roles/jobs_cancel/tasks/main.yml:6]: i think its good way to indicate private vars.

**Review by kedark3 — APPROVED** [review]
Overall looks great at a quick glance. I will let Tom finalize it as its  big PR :)

File: roles/jobs_cancel/tasks/main.yml (hunk 14 above)
Comment (Tompage1994) [reply roles/jobs_cancel/tasks/main.yml:6]: Im happy with it, we should probably make sure we have one standard and go back and change any which don't match

File: roles/project_update/README.md (hunk 15 above)
Comment (Tompage1994) [reply roles/project_update/README.md:46]: I think having just one copy of the list of projects is a good idea. But under this way we dont have any logic to say whether we actually want to update or not. Maybe an option of `update: true` would be a good idea and we can just do a `when: __project_update_update_item.update`

**Review by Tompage1994 — APPROVED** [review]
Looks good now!

Thanks for the HUGE load of fixes

**Review by Tompage1994 — CHANGES_REQUESTED** [review]
Actually, can you just add a changelog for this addition as it does add new functionality. Then happy to merge

File: roles/jobs_cancel/tasks/main.yml (hunk 14 above)
Comment (sean-m-sullivan) [reply roles/jobs_cancel/tasks/main.yml:6]: I'm thinking the same Tom, We adopt more of the Oasis standards, and this one I like, and we'll make an issue and go back and change the others.

File: roles/project_update/README.md (hunk 15 above)
Comment (sean-m-sullivan) [reply roles/project_update/README.md:46]: Testing that now, think its a good solution

Comment (sean-m-sullivan) [reply roles/project_update/README.md:46]: Done and tested.

**Pull request merged**